### PR TITLE
READY | alleviating browser throttling on WebRTC transport

### PR DESCRIPTION
**Pull request by ya7ya**
This PR is related to https://github.com/ipfs/js-ipfs/issues/1019 and https://github.com/ipfs/js-ipfs-bitswap/pull/152 , it basically implements `MAX_MESSAGE_SIZE` on the webRTC transport to avoid being throttled by the browser.

however as you can see, it fails one of the tests. I would appreciate some help in debugging this since It's been frustrating.

I've tried to use `pull-block` but it didn't work (check the commented out code in `dial`)

**note** this also still requires some clean-up as you might have noticed 😄 

cc. @diasdavid

### CODE REVIEW

**Comment by pgte** [comment]
Looks like there's a race condition when ending the stream.
Perhaps you could try using the `pump` package to pipe the channel into block-stream and back, perhaps it will help you avoiding or finding the source of the issue?

**Comment by ya7ya** [comment]
@pgte Thanks for the `pump` package recommendation 👍 . it worked like a charm. no more race condition.

The tests are passing locally just fine. let me know there are any edits required 😄

**Comment by ya7ya** [comment]
BTW. the tests are passing successfully on my machine, I reckon it has to do with the `wrtc` and `electron-wrtc` build issues on linux mentioned [here](https://github.com/ipfs/js-ipfs#is-there-webrtc-support-for-js-ipfs-with-nodejs)

testnodejs

File: src/index.js [hunk 1]
```diff
@@ -69,7 +73,14 @@ class WebRTCStar {
 
     const channel = new SimplePeer(spOptions)
 
-    const conn = new Connection(toPull.duplex(channel))
+    const block = new BlockStream(this.blockStreamSize, {nopad: true})
+
+    pump(channel, block, channel, (err) => {
+      if (err) return callback(err)
```
Comment (pgte) [inline src/index.js]: I see a possible double-callback here when an error occurs. Maybe use the `once` package to make sure you only call the callback once?

File: src/index.js [hunk 2]
```diff
@@ -169,7 +180,13 @@ class WebRTCStar {
 
         const channel = new SimplePeer(spOptions)
 
-        const conn = new Connection(toPull.duplex(channel))
+        const block = new BlockStream(this.blockStreamSize, {nopad: true})
+
+        pump(channel, block, channel, (err) => {
+          if (err) return callback(err)
```
Comment (pgte) [inline src/index.js]: same here

File: src/index.js (hunk 1 above)
Comment (ya7ya) [reply src/index.js]: @pgte , once is already used few lines before https://github.com/libp2p/js-libp2p-webrtc-star/pull/145/files/1b0d1ed130d6d447c3d0fc54b0abee13fdde11e7#diff-1fdf421c05c1140f6d71444ea2b27638R62

```javascript
callback = callback ? once(callback) : noop
// ....
// ....
const block = new BlockStream(this.blockStreamSize, {nopad: true})

pump(channel, block, channel, (err) => {
  if (err) return callback(err)
})
```

Does that cover it or am i missing something here ?

**Comment by ya7ya** [comment]
I edited the commit msgs that weren't standard. so I had to `push --force` this , sorry about that 😅

File: src/index.js (hunk 1 above)
Comment (pgte) [reply src/index.js]: Ha, it is, I missed it, sorry. Looks good to me!

File: src/index.js (hunk 2 above)
Comment (pgte) [reply src/index.js]: nvm

**Pull request merged**